workflow/scripts/rNA_flowcells.Rmd

```diff
@@ -353,21 +347,70 @@ cgroups <- as.factor(multiQC$flowcell_lanes)
 cpalette <- brewer.pal(nlevels(cgroups), "Paired")
 
-p <- plot_ly(shared_metadata,
-  x = ~PC1, y = ~PC2, z = ~PC3, color = cgroups, colors = cpalette, hoverinfo = "text", marker = list(size = 8),
-  text = ~ paste(
-    "</br><b> Sample: ", Sample, "</br> Flowcell Lanes: ", flowcell_lanes, "</br> medTIN: ", median_tin, "</b></br></br> % Aligned: ", percent_aligned,
-    "</br> % Dup: ", percent_duplication, "</br> % Coding: ", pct_coding_bases, "</br> % UTR: ", pct_utr_bases,
-    "</br> % Intronic: ", pct_intronic_bases, "</br></br> Sequence Range: ", sequence_length,
-    "</br> GC Content: ", gc_content, "</br> Inner Distance Maxima: ", inner_distance_maxima, "</br> Insert Size: ", median_insert_size
-  )
-) %>%
-  add_markers() %>%
-  layout(scene = list(
-    xaxis = list(title = paste0("PC1 (", pc1, "%)")),
-    yaxis = list(title = paste0("PC2 (", pc2, "%)")),
-    zaxis = list(title = paste0("PC3 (", pc3, "%)"))
-  ))
+# if project has low sample differential
+if (is.na(pc3)) {
+  p <- plot_ly(shared_metadata,
+    x = ~PC1,
+    y = ~PC2,
+    color = cgroups,
+    colors = cpalette,
+    hoverinfo = "text",
+    marker = list(size = 8),
+    text = ~ paste(
+      "</br><b> Sample: ",
+      Sample, "</br> Flowcell Lanes: ",
+      flowcell_lanes, "</br> medTIN: ",
+      median_tin, "</b></br></br> % Aligned: ",
+      percent_aligned, "</br> % Dup: ",
+      percent_duplication, "</br> % Coding: ",
+      pct_coding_bases, "</br> % UTR: ",
+      pct_utr_bases,
+      "</br> % Intronic: ",
+      pct_intronic_bases, "</br></br> Sequence Range: ",
+      sequence_length,
+      "</br> GC Content: ",
+      gc_content, "</br> Inner Distance Maxima: ",
+      inner_distance_maxima, "</br> Insert Size: ", median_insert_size
+    )
+  ) %>%
+    add_markers() %>%
+    layout(scene = list(
+      xaxis = list(title = paste0("PC1 (", pc1, "%)")),
+      yaxis = list(title = paste0("PC2 (", pc2, "%)"))
+    ))
+} else {
+  p <- plot_ly(shared_metadata,
+    x = ~PC1,
+    y = ~PC2,
+    z = ~PC3, color = cgroups,
+    colors = cpalette,
+    hoverinfo = "text",
+    marker = list(size = 8),
+    text = ~ paste(
+      "</br><b> Sample: ",
+      Sample, "</br> Flowcell Lanes: ",
+      flowcell_lanes, "</br> medTIN: ",
+      median_tin, "</b></br></br> % Aligned: ",
+      percent_aligned, "</br> % Dup: ",
+      percent_duplication, "</br> % Coding: ",
+      pct_coding_bases, "</br> % UTR: ",
+      pct_utr_bases,
+      "</br> % Intronic: ",
+      pct_intronic_bases, "</br></br> Sequence Range: ",
+      sequence_length,
+      "</br> GC Content: ",
+      gc_content, "</br> Inner Distance Maxima: ",
+      inner_distance_maxima, "</br> Insert Size: ", median_insert_size
+    )
+  ) %>%
+    add_markers() %>%
+    layout(scene = list(
+      xaxis = list(title = paste0("PC1 (", pc1, "%)")),
+      yaxis = list(title = paste0("PC2 (", pc2, "%)")),
+      zaxis = list(title = paste0("PC3 (", pc3, "%)"))
+    ))
+}
+
```

Comment:
I think we should write a function to avoid having nearly the same code twice. We can use argument unpacking to pass the max number of PCs to plotly. Here's a small example showing the design I have in mind:

```r
plot_pca_expression <- function(dat, pca_max) {
  if (pca_max > 3 | pca_max < 2) {
    stop(glue::glue("Number of principle components must be 2 or 3, but is {pca_max}"))
  }
  pcs <- c(x = ~PC1, y = ~PC2, z = ~PC3)[1:pca_max]
  args <-c(pcs, list(
    data = dat,
    hoverinfo = "text",
    marker = list(size = 8)
    # additional named arguments to `plot_ly` go here
  ))
  do.call(plot_ly, args) %>%
    add_markers()
    # additional piped functions after `plot_ly` go here
}

dat <- data.frame(PC1 = runif(10), PC2 = runif(10), PC3 = runif(10))
pca_max <- 2
plot_pca_expression(dat, pca_max)

pca_max <- 3
plot_pca_expression(dat, pca_max)
```

Reply:
take a look at my latest push - created a function to handle both of these. test run was successful.